File: src/main/java/me/frauenfelderflorian/worldutils/commands/CSettings.java
package me.frauenfelderflorian.worldutils.commands;

import me.frauenfelderflorian.worldutils.Messages;
import me.frauenfelderflorian.worldutils.WorldUtils;
import me.frauenfelderflorian.worldutils.config.Prefs;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabExecutor;
import org.bukkit.entity.Player;
import org.bukkit.util.StringUtil;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

/**
 * CommandExecutor and TabCompleter for command settings
 */
public record CSettings(WorldUtils plugin) implements TabExecutor {
    public static final String CMD = "settings";

    /**
     * Done when command sent
     *
     * @param sender  sender of the command
     * @param command sent command
     * @param alias   used alias
     * @param args    used arguments
     * @return true if correct command syntax used and no errors, false otherwise
     */
    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String alias,
                             String[] args) {
        if (args.length == 3) {
            //correct number of arguments
            Prefs.Option setting = Prefs.Option.get(args[0], args[1]);
            if (setting != null && List.of("true", "false", "null").contains(args[2])) {
                //correct setting and value
                if (!setting.isGlobal() && sender instanceof Player) {
                    //personal setting
                    switch (args[2]) {
                        case "true" -> {
                            //set to true
                            plugin.prefs.set((Player) sender, setting, true, true);
                            Messages.sendMessage(sender, Messages.settingSet(setting, "true"));
                            return true;
                        }
                        case "false" -> {
                            //set to false
                            plugin.prefs.set((Player) sender, setting, false, true);
                            Messages.sendMessage(sender, Messages.settingSet(setting, "false"));
                            return true;
                        }
                        case "null" -> {
                            //remove setting
                            plugin.prefs.remove((Player) sender, setting);
                            Messages.sendMessage(sender, Messages.settingSet(setting, "null"));
                            Messages.sendMessage(sender, "§cUse with caution: §oThe plugin might not work correctly!");
                            return true;
                        }
                    }
                } else if (setting.isGlobal()
                        && (sender.isOp() || !plugin.prefs.getBoolean(Prefs.Option.SETTINGS_NEED_OP))) {
                    //global setting
                    switch (args[2]) {
                        case "true" -> {
                            //set to true
                            plugin.prefs.set(setting, true, true);
                            Messages.sendMessage(Messages.settingSet(setting, "true"));
                            return true;
                        }
                        case "false" -> {
                            //set to false
                            plugin.prefs.set(setting, false, true);
                            Messages.sendMessage(sender, Messages.settingSet(setting, "false"));
                            return true;
                        }
                        case "null" -> {
                            //remove setting
                            plugin.prefs.remove(setting);
                            Messages.sendMessage(Messages.settingSet(setting, "null"));
                            Messages.sendMessage("§cUse with caution: §oThe plugin might not work correctly!");
                            return true;
                        }
                    }
                }
            } else Messages.wrongArguments(sender);
        } else Messages.wrongArgumentNumber(sender);
        return false;
    }

    /**
     * Done while entering command
     *
     * @param sender  sender of the command
     * @param command sent command
     * @param alias   used alias
     * @param args    used arguments
     * @return List of Strings for tab completion
     */
    @Override
    public List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command command, @NotNull String alias,
                                      String[] args) {
        List<String> completions = new ArrayList<>();
        switch (args.length) {
            case 1 -> {
                //command being entered
                if (sender.isOp() || !plugin.prefs.getBoolean(Prefs.Option.SETTINGS_NEED_OP))
                    StringUtil.copyPartialMatches(args[0], Prefs.Option.getCommands(true), completions);
                if (sender instanceof Player)
                    StringUtil.copyPartialMatches(args[0], Prefs.Option.getCommands(false), completions);
            }
            case 2 -> {
                //setting being entered
                if (Prefs.Option.getCommands().contains(args[0])) {
                    if (sender.isOp() || !plugin.prefs.getBoolean(Prefs.Option.SETTINGS_NEED_OP))
                        StringUtil.copyPartialMatches(args[1], Prefs.Option.getSettings(args[0], true),
                                completions);
                    if (sender instanceof Player)
                        StringUtil.copyPartialMatches(args[1], Prefs.Option.getSettings(args[0], false),
                                completions);
                }
            }
            case 3 -> {
                //value being entered
                Prefs.Option setting = Prefs.Option.get(args[0], args[1]);
                if (setting != null && (sender instanceof Player && !setting.isGlobal()
                        || (sender.isOp() || !plugin.prefs.getBoolean(Prefs.Option.SETTINGS_NEED_OP))
                        && setting.isGlobal()))
                    StringUtil.copyPartialMatches(args[2], List.of("true", "false"), completions);
            }
        }
        return completions;
    }
}

File: src/main/java/me/frauenfelderflorian/worldutils/commands/CSettings.kt
package me.frauenfelderflorian.worldutils.commands

import me.frauenfelderflorian.worldutils.*
import me.frauenfelderflorian.worldutils.config.Option
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.command.TabExecutor
import org.bukkit.entity.Player
import org.bukkit.util.StringUtil

/** [TabExecutor] for command settings */
class CSettings(val plugin: WorldUtils) : TabExecutor {
    /** Done when the [sender] sends the registered [command] with [args] */
    override fun onCommand(
        sender: CommandSender, command: Command, alias: String, args: Array<String>,
    ): Boolean {
        if (args.size == 3) {
            //correct number of arguments
            val setting = Option[args[0], args[1]]
            if (setting != null && args[2] in listOf("true", "false", "null")) {
                //correct setting and value
                if (!setting.isGlobal && sender is Player) {
                    //personal setting
                    when (args[2]) {
                        "true" -> {
                            //set to true
                            plugin.prefs.set(
                                player = sender, setting = setting, value = true, log = true
                            )
                            sendMessage(plugin, sender, settingSetMessage(setting, "true"))
                            return true
                        }

                        "false" -> {
                            //set to false
                            plugin.prefs.set(
                                player = sender, setting = setting, value = false, log = true
                            )
                            sendMessage(plugin, sender, settingSetMessage(setting, "false"))
                            return true
                        }

                        "null" -> {
                            //remove setting
                            plugin.prefs.remove(sender, setting)
                            sendMessage(plugin, sender, settingSetMessage(setting, "null"))
                            sendMessage(
                                plugin,
                                sender,
                                "§cUse with caution: §oThe plugin might not work correctly!"
                            )
                            return true
                        }
                    }
                } else if (setting.isGlobal && (sender.isOp || !plugin.prefs.getBoolean(Option.SETTINGS_NEED_OP))) {
                    //global setting
                    when (args[2]) {
                        "true" -> {
                            //set to true
                            plugin.prefs.set(setting = setting, value = true, log = true)
                            sendMessage(plugin, settingSetMessage(setting, "true"))
                            return true
                        }

                        "false" -> {
                            //set to false
                            plugin.prefs.set(setting = setting, value = false, log = true)
                            sendMessage(plugin, sender, settingSetMessage(setting, "false"))
                            return true
                        }

                        "null" -> {
                            //remove setting
                            plugin.prefs.remove(setting)
                            sendMessage(plugin, settingSetMessage(setting, "null"))
                            sendMessage(
                                plugin, "§cUse with caution: §oThe plugin might not work correctly!"
                            )
                            return true
                        }
                    }
                }
            } else {
                wrongArguments(plugin, sender)
            }
        } else {
            wrongArgumentNumber(plugin, sender)
        }
        return false
    }

    /**
     * Done while the [sender] enters the registered [command] with [args]
     *
     * @return List of Strings for tab completion
     */
    override fun onTabComplete(
        sender: CommandSender, command: Command, alias: String, args: Array<String>,
    ): List<String> {
        val completions = mutableListOf<String>()
        when (args.size) {
            1 -> {
                //command being entered
                if (sender.isOp || !plugin.prefs.getBoolean(Option.SETTINGS_NEED_OP)) {
                    StringUtil.copyPartialMatches(args[0], Option.getCommands(true), completions)
                }
                if (sender is Player) {
                    StringUtil.copyPartialMatches(args[0], Option.getCommands(false), completions)
                }
            }

            2 -> {
                //setting being entered
                if (Option.commands.contains(args[0])) {
                    if (sender.isOp || !plugin.prefs.getBoolean(Option.SETTINGS_NEED_OP)) {
                        StringUtil.copyPartialMatches(
                            args[1], Option.getSettings(args[0], true), completions
                        )
                    }
                    if (sender is Player) {
                        StringUtil.copyPartialMatches(
                            args[1], Option.getSettings(args[0], false), completions
                        )
                    }
                }
            }

            3 -> {
                //value being entered
                val setting = Option[args[0], args[1]]
                if (setting != null && (sender is Player && !setting.isGlobal || (sender.isOp || !plugin.prefs.getBoolean(
                        Option.SETTINGS_NEED_OP
                    )) && setting.isGlobal)
                ) {
                    StringUtil.copyPartialMatches(args[2], listOf("true", "false"), completions)
                }
            }
        }
        return completions
    }

    companion object {
        const val CMD = "settings"
    }
}
